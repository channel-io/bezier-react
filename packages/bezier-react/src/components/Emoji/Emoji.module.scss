@use '../../styles/mixins/dimension';

$emoji-sizes: '16', '20', '24', '30', '36', '42', '48', '60', '72', '90', '120';

@mixin emoji-background($emoji-size) {
  background-image: var(--b-emoji-background-image);
  background-position: center;
  background-repeat: no-repeat;
  background-size: $emoji-size $emoji-size;
  background-clip: content-box;
}

.Emoji {
  --b-emoji-size: initial;

  position: relative;
  box-sizing: content-box;
  display: flex;
  flex-shrink: 0;
  align-items: center;
  justify-content: center;
  @include dimension.square(var(--b-emoji-size));
  @include emoji-background(var(--b-emoji-size));

  @each $size in $emoji-sizes {
    &:where(.size-#{$size}) {
      --b-emoji-size: #{$size}px;
    }
  }
}
