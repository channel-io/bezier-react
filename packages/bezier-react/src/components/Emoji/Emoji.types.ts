import type { BezierComponentProps, SizeProps } from '~/src/types/props'

export type EmojiSize =
  | '16'
  | '20'
  | '24'
  | '30'
  | '36'
  | '42'
  | '48'
  | '60'
  | '72'
  | '90'
  | '120'

interface EmojiOwnProps {
  /**
   * Name of the emoji. e.g. 'grinning', 'smiley', etc.
   */
  name: string
  /**
   * @deprecated
   * `imageUrl` is created in the component and will be removed in the next major version.
   */
  imageUrl?: string
}

export interface EmojiProps
  extends BezierComponentProps<'div'>,
    SizeProps<EmojiSize>,
    EmojiOwnProps {}
