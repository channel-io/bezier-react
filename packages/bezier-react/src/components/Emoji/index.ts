export { Emoji } from './Emoji'
export type {
  EmojiProps,
  EmojiSize,
} from './Emoji.types'
