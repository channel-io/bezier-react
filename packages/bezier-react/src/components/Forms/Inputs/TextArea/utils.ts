import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
