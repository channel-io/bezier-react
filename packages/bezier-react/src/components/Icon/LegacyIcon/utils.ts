/* Internal Dependencies */
// eslint-disable-next-line no-restricted-imports
import icons, { type IconName } from '../generated'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { Icon, AllIcon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-react'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(<svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
