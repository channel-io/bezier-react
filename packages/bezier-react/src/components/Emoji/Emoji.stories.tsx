import {
  type Meta,
  type StoryObj,
} from '@storybook/react'

import { Emoji } from './Emoji'
import { type EmojiProps } from './Emoji.types'

const MOCK_EMOJI_URL = 'https://cf.exp.channel.io/asset/emoji/images/80/blush.png'

const meta: Meta<typeof Emoji> = {
  component: Emoji,
  argTypes: {
    imageUrl: {
      control: {
        type: 'text',
      },
    },
  },
}

export default meta

export const Primary: StoryObj<EmojiProps> = {
  args: {
    size: '24',
    imageUrl: MOCK_EMOJI_URL,
  },
}

