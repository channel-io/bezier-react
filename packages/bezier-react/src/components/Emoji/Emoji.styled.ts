import {
  centeredBackgroundImage,
  styled,
} from '~/src/foundation'

import { type InterpolationProps } from '~/src/types/Foundation'

import { type EmojiSize } from './Emoji.types'

interface IconProps extends InterpolationProps {
  size: EmojiSize
  imageUrl: string
}

export const Icon = styled.div<IconProps>`
  position: relative;
  box-sizing: content-box;
  display: flex;
  flex-shrink: 0;
  align-items: center;
  justify-content: center;
  width: ${({ size }) => size}px;
  height: ${({ size }) => size}px;

  ${({ size }) => centeredBackgroundImage<EmojiSize>(size)}

  ${({ interpolation }) => interpolation}
`
