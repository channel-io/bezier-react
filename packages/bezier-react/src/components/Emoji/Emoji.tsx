import React, { type CSSProperties, forwardRef } from 'react'

import classNames from 'classnames'

import { isDev } from '~/src/utils/assert'
import { cssUrl } from '~/src/utils/style'

import { type EmojiProps } from './Emoji.types'

import styles from './Emoji.module.scss'

export const EMOJI_TEST_ID = 'bezier-emoji'

const getEmojiUrl = (name: EmojiProps['name'], size: '160' | '80' | '44') => {
  return `https://cf${isDev() ? '.exp' : ''}.channel.io/asset/emoji/images/${size}/${encodeURIComponent(name)}.png`
}

/**
 * `Emoji` is a component for representing emoji with variant size.
 * @example
 * ```tsx
 * <Emoji
 *   name="A"
 *   size="20"
 * />
 * ```
 */
export const Emoji = forwardRef<HTMLDivElement, EmojiProps>(function Emoji(
  { style, imageUrl, className, name, size = '24', ...rest },
  forwardedRef
) {
  const assetSize = Number(size) >= 60 ? '160' : '80'

  return (
    <div
      ref={forwardedRef}
      role="img"
      aria-description={name}
      style={
        {
          '--b-emoji-background-image': cssUrl(
            imageUrl ?? getEmojiUrl(name, assetSize)
          ),
          ...style,
        } as CSSProperties
      }
      className={classNames(styles.Emoji, styles[`size-${size}`], className)}
      data-testid={EMOJI_TEST_ID}
      {...rest}
    />
  )
})
