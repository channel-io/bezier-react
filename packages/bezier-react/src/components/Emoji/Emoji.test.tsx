/* External dependencies */
import React from 'react'

/* Interanal dependencies */
import { render } from '~/src/utils/testUtils'
import type EmojiProps from './Emoji.types'
import Emoji, { EMOJI_TEST_ID } from './Emoji'
import { EmojiSize } from './Emoji.types'

const MOCK_EMOJI_URL = 'https://cf.exp.channel.io/asset/emoji/images/80/a.png'

describe('Emoji Test >', () => {
  const defaultProps: EmojiProps = { imageUrl: MOCK_EMOJI_URL, name: 'a' }

  const renderComponent = (props?: Partial<EmojiProps>) => render(
    <Emoji {...defaultProps} {...props} />,
  )

  it('snapshot', () => {
    const { getByTestId } = renderComponent()
    const emoji = getByTestId(EMOJI_TEST_ID)

    expect(emoji).toMatchSnapshot()
  })

  it('should render default style', () => {
    const { getByTestId } = renderComponent()
    const emoji = getByTestId(EMOJI_TEST_ID)

    expect(emoji).toHaveStyle('position: relative')
    expect(emoji).toHaveStyle('box-sizing: content-box')
    expect(emoji).toHaveStyle('display: flex')
    expect(emoji).toHaveStyle('flex-shrink: 0')
    expect(emoji).toHaveStyle('align-items: center')
    expect(emoji).toHaveStyle('justify-content: center')
  })

  it('should render with emoji image', () => {
    const { getByTestId } = renderComponent()
    const emoji = getByTestId(EMOJI_TEST_ID)

    expect(emoji).toHaveStyle('background-image: var(--background-image)')
  })

  it('should render emoji icon with size prop - size16', () => {
    const { getByTestId } = renderComponent({ size: EmojiSize.Size16 })
    const emoji = getByTestId(EMOJI_TEST_ID)

    expect(emoji).toHaveStyle('width: 16px')
    expect(emoji).toHaveStyle('height: 16px')
  })

  it('should render emoji icon with size prop - size42', () => {
    const { getByTestId } = renderComponent({ size: EmojiSize.Size42 })
    const emoji = getByTestId(EMOJI_TEST_ID)

    expect(emoji).toHaveStyle('width: 42px')
    expect(emoji).toHaveStyle('height: 42px')
  })

  it('should render emoji icon with size prop - size120', () => {
    const { getByTestId } = renderComponent({ size: EmojiSize.Size120 })
    const emoji = getByTestId(EMOJI_TEST_ID)

    expect(emoji).toHaveStyle('width: 120px')
    expect(emoji).toHaveStyle('height: 120px')
  })
})
