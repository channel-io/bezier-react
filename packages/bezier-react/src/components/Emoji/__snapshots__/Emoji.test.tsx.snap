// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emoji Test > snapshot 1`] = `
<div
  aria-label="a"
  class="Emoji size-24"
  data-testid="bezier-emoji"
  role="img"
  style="--b-emoji-background-image: url(https://cf.exp.channel.io/asset/emoji/images/80/a.png);"
/>
`;
